File: framework/languages/dotnet/e2e/SpotService/Client/Scenarios/SmD12Scenario.cs
using SpotService.Client;
using SpotService.Shared;
using Systems.Zlink.Stream.Connector.Contracts;

namespace SpotService.Client.Scenarios;

internal static class SmD12Scenario
{
    public static async Task RunAsync(string sessionAStreamEndpoint, string sessionBStreamEndpoint)
    {
        const string actorId = "actor-sm-d12-transfer";

        await using var first = await ConnectAndAuthWithRetryAsync(
            sessionAStreamEndpoint,
            new AuthReq(actorId, "api transfer", "play-a"));
        var firstReply = await first.Request(new ActorPingReq("before-transfer"))
            .PacketName("ActorPingReq")
            .Async<ActorPingReply>();
        ScenarioAssert.That(firstReply.ActorId == actorId, "SM-D12 first api actor mismatch.");
        ScenarioAssert.That(firstReply.NodeRid == "play-a", "SM-D12 first api node mismatch.");
        ScenarioAssert.That(firstReply.Seen == 1, "SM-D12 expected initial actor state.");
        await first.Close.Async();

        await using var second = await ConnectAndAuthWithRetryAsync(
            sessionBStreamEndpoint,
            new AuthReq(actorId, "api transfer", "play-a"));
        var snapshot = await second.Request(new SnapshotReq(actorId))
            .PacketName("SnapshotReq")
            .Async<SnapshotReply>();
        ScenarioAssert.That(snapshot.ActorId == actorId, "SM-D12 snapshot actor mismatch.");
        ScenarioAssert.That(snapshot.Seen == 1, "SM-D12 actor state was not preserved across apis.");

        var pushed = second.WaitFor<ActorPushNotify>().Async().AsTask();
        var resumed = await second.Request(new ActorPushReq("after-transfer"))
            .PacketName("ActorPushReq")
            .Async<ActorPingReply>();
        var notify = await pushed;
        ScenarioAssert.That(resumed.ActorId == actorId, "SM-D12 resumed actor mismatch.");
        ScenarioAssert.That(resumed.NodeRid == "play-a", "SM-D12 resumed node mismatch.");
        ScenarioAssert.That(resumed.Seen == 2, "SM-D12 resumed actor state mismatch.");
        ScenarioAssert.That(notify.Payload.ActorId == actorId, "SM-D12 resumed push actor mismatch.");
        ScenarioAssert.That(notify.Payload.Value == "after-transfer", "SM-D12 resumed push value mismatch.");
        ScenarioAssert.That(notify.Payload.Seen == 2, "SM-D12 resumed push state mismatch.");

        Console.WriteLine("operation SpotService.sm-d12 passed");
    }

    static async Task<IZlinkStreamConnector> ConnectAndAuthWithRetryAsync(string endpoint, AuthReq auth)
    {
        var deadline = DateTimeOffset.UtcNow + TimeSpan.FromSeconds(10);
        Exception? last = null;
        while (DateTimeOffset.UtcNow < deadline)
        {
            var client = CreateClient(endpoint);
            try
            {
                await client.Connect.Async();
                await client.Request(auth)
                    .PacketName("AuthReq")
                    .Async<AuthReply>();
                return client;
            }
            catch (Exception ex) when (ex is ZlinkStreamException or TimeoutException)
            {
                last = ex;
                await client.DisposeAsync();
                await Task.Delay(200);
            }
        }

        throw new InvalidOperationException(
            last is null ? $"Actor auth did not become routable: {auth.ActorId}" : $"Actor auth did not become routable: {auth.ActorId}. Last error: {last.Message}",
            last);
    }

    static IZlinkStreamConnector CreateClient(string endpoint)
    {
        ScenarioAssert.That(!string.IsNullOrWhiteSpace(endpoint), "stream endpoint is required.");
        return ZlinkStreamConnectorFactory.Create(new ZlinkStreamConnectorOptions
        {
            Endpoint = new Uri(endpoint),
            ConnectTimeout = TimeSpan.FromSeconds(5),
            RequestTimeout = TimeSpan.FromSeconds(5),
            Heartbeat = new ZlinkStreamHeartbeatOptions { Enabled = false },
            DispatchMode = ZlinkStreamDispatchMode.Immediate,
            MaxReceivedMessages = 1024,
        });
    }
}

File: framework/languages/node/e2e/SpotService/Client/Scenarios/sm-d12-scenario.ts
import {
  zlinkStreamConnectorFactory,
  zlinkStreamJsonCodec,
  ZlinkStreamDispatchMode
} from '@zlink-systems/stream-connector';
import type {
  ActorPingRes,
  ActorPingReq,
  ActorPushNotify,
  ActorPushReq,
  AuthRes,
  AuthReq,
  SnapshotRes,
  SnapshotReq
} from '../../Shared/messages';
import type { ClientOptions } from '../Support/client-options';
import { ensure } from '../Support/scenario-assert';
import { decodeStreamReply } from '../Support/stream-reply';

export async function runSmD12(options: ClientOptions): Promise<void> {
  const actorId = 'actor-sm-d12-transfer';
  const first = createStreamClient(options.sessionAStreamEndpoint);
  let second: ReturnType<typeof createStreamClient> | undefined;

  await first.connect();
  try {
    await first
      .request({
        actorId,
        displayName: 'api transfer',
        nodeRid: 'play-a'
      } satisfies AuthReq)
      .packetName('AuthReq')
      .timeout(5000)
      .submit<AuthRes>();

    const firstReply = decodeStreamReply<ActorPingRes>(await first
      .request({ value: 'before-transfer' } satisfies ActorPingReq)
      .packetName('ActorPingReq')
      .timeout(5000)
      .submit());
    ensure(firstReply.actorId === actorId, 'SM-D12 first api actor mismatch.');
    ensure(firstReply.nodeRid === 'play-a', 'SM-D12 first api node mismatch.');
    ensure(firstReply.seen === 1, 'SM-D12 expected initial actor state.');

    await first.close();

    second = createStreamClient(options.sessionBStreamEndpoint);
    await second.connect();
    await second
      .request({
        actorId,
        displayName: 'api transfer',
        nodeRid: 'play-a'
      } satisfies AuthReq)
      .packetName('AuthReq')
      .timeout(5000)
      .submit<AuthRes>();

    const snapshot = decodeStreamReply<SnapshotRes>(await second
      .request({ actorId } satisfies SnapshotReq)
      .packetName('SnapshotReq')
      .timeout(5000)
      .submit());
    ensure(snapshot.actorId === actorId, 'SM-D12 snapshot actor mismatch.');
    ensure(snapshot.seen === 1, 'SM-D12 actor state was not preserved across apis.');

    const pushed = second.waitFor<ActorPushNotify>('ActorPushNotify')
      .where((message) => message.payload.actorId === actorId)
      .timeout(10000)
      .submit();
    const resumed = decodeStreamReply<ActorPingRes>(await second
      .request({ value: 'after-transfer' } satisfies ActorPushReq)
      .packetName('ActorPushReq')
      .timeout(5000)
      .submit());
    const notify = await pushed;
    ensure(resumed.actorId === actorId, 'SM-D12 resumed actor mismatch.');
    ensure(resumed.nodeRid === 'play-a', 'SM-D12 resumed node mismatch.');
    ensure(resumed.seen === 2, 'SM-D12 resumed actor state mismatch.');
    ensure(notify.payload.actorId === actorId, 'SM-D12 resumed push actor mismatch.');
    ensure(notify.payload.value === 'after-transfer', 'SM-D12 resumed push value mismatch.');
    ensure(notify.payload.seen === 2, 'SM-D12 resumed push state mismatch.');
  } finally {
    await first.close().catch(() => undefined);
    await second?.close().catch(() => undefined);
  }

  console.log('scenario SM-D12 passed');
}

function createStreamClient(endpoint: string) {
  return zlinkStreamConnectorFactory.create({
    endpoint,
    codec: zlinkStreamJsonCodec,
    dispatchMode: ZlinkStreamDispatchMode.Immediate,
    heartbeat: { enabled: false },
    maxReceivedMessages: 1024,
    waitTimeoutMs: 10000
  });
}
